Sources/LanguageServerProtocol/Requests/MacroExpansionRequest.swift

```diff
@@ -0,0 +1,39 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Request the expansion of the macro at a given use site.
+/// **LSP Extension**.
```

Comment:
I like the idea of having a URL scheme to describe the expanded macro buffers and implementing a `sourcekit-lsp/getReferenceDocument` request to get the generated contents to the client. The main advantage I see with it are:
- We have a URI that we can use to talk about the expanded macro buffer. Adding support for semantic functionality within the generated buffer is just a matter of implementing LSP requests for that URL scheme.
- We can use it to show the content of generated interfaces where we currently write the generated interface to disk and then open it.

I don’t quite understand where the `sourcekit-lsp/peekDocument` request comes in. My design of this feature would be using `ExecuteCommandRequest`.

1. The client sends a `textDocument/codeAction` request to the server
2. If we are at a macro, the server responds with a `Expand Macro` code action that has a command with the custom URI scheme in the `arguments`.
3. If the user decides to pick that action, the client sends a `workspace/executeCommand` for that command.
4. And now is where it get’s a little tricky: The server replies with some kind of special response (reply of `workspace/executeCommand` is `LSPAny?`) that the client understands as: OK, I need to show an inline text buffer with the text of this URI scheme. E.g. the reply is 
```json
{ 
}
```

Reply:
Yes, your `showInlineBuffer` reply is effectively what I meant by `sourcekit-lsp/peekDocument` (only that it wouldn't be a reply, but a separate request (which would effectively let the server invoke it multiple times, if it so desires).

But until upstream has decided on a specific design, I would be fine either way.

Reply:
> Is peekDocument a request you came up with or is it based on some proposal?

The proposal specifies `openDocument`, which effectively falls in a similar vein, but doesn't suit our use case as well (if I understand it correctly, it would allow the server to open documents in the user's editor), therefore `peekDocument` was [my suggestion](https://github.com/microsoft/language-server-protocol/issues/336#issuecomment-1761708964) for a command that would open an inline document.

> In my mind the client should always be driving the user interaction and the LSP server should just reply to the user’s request, that’s why I think peekDocument would be a little weird.

I agree in principle, the only neat thing I could see about the `peekDocument` request is that the client wouldn't have to implement a lot of server-specific logic, once/if standardized (i.e. it would only have to invoke the command and the rest is covered by standard LSP requests). I also agree that it does feel a bit "imperative" to have the server perform actions on the client as opposed to the rather declarative model of e.g. just querying for macro expansions.

Perhaps the best solution would be to keep our non-standard macro expansion request for now and simply return a `swift-macro-expansion://` URI which the client then resolves via `sourcekit-lsp/getReferenceDocument` (the part of the proposal that I consider the most likely to be implemented in this form anyway)? This way we could handle nested expansions, while keeping the client-driven model, plus we wouldn't have to assign a new meaning to `executeCommand` responses. Unless that response is standardized as well, we'd need custom logic on the client one way or the other.

What do you think?

Reply:
I do see the advantage in being able to return macro expansions as a code action though... perhaps your design is the way to go then?

Reply:
> and simply return a `swift-macro-expansion://` URI

Which request would this be a reply to (and in which field)? AFAICT `CodeAction` doesn’t have a URI field (unless we just shove it in the `arguments: [LSPAny]?`)

> Edit: I am not sure if there's a way to handle custom responses for commands triggered by a code action though, since code actions are handled by the `vscode-languageclient` library.

Oh, that would indeed be a major limitation, unfortunately.

Reply:
> keep our non-standard macro expansion request for now and simply return a `swift-macro-expansion://`

with that I meant using the current design from this PR and return something along the lines of 

```swift
struct MacroExpansion {
  let uri: DocumentURI // = DocumentURI("swift-macro-expansion://...")
  let position: Position
}
```

from `sourcekit-lsp/macroExpansion`, which the client then resolves. But, yeah that wouldn't work with as a code action, unfortunately, the client would have to implement the command manually (as my `vscode-swift` PR currently does).

Reply:
> But, yeah that wouldn't work as a code action, unfortunately, the client would have to implement the command manually (as my `vscode-swift` PR currently does).

That’s the major downside I see with the current design. IIUC it currently offers “Expand Macro” everywhere, even when the cursor isn’t currently on a macro. I think implementing it on top of Code Actions would be the cleanest design that also yields the best UI results.

Reply:
So implementing it as a command where the server calls the client back seems like the only option that fulfills all requirements (works as a code action and doesn't need a custom response)?

Reply:
Yes, I think that is the case. So your design was the right one all along 😄